File: src/emu/cpu_dis.js
import * as opcode from "./cpu_opcodes.js"


export class Disassembler
{
	static disassembleInstruction(addr, byte1, byte2, byte3)
	{
		switch (byte1)
		{
			case opcode.BRK    : return instrIMP("brk", byte2, byte3)

			case opcode.PLA    : return instrIMP("pla", byte2, byte3)
			case opcode.PLP    : return instrIMP("plp", byte2, byte3)
			case opcode.PHA    : return instrIMP("pha", byte2, byte3)
			case opcode.PHP    : return instrIMP("php", byte2, byte3)

			case opcode.JMP_ABS: return instrABS("jmp", byte2, byte3)
			case opcode.JMP_IND: return instrIND("jmp", byte2, byte3)
			case opcode.BPL    : return instrREL("bpl", byte2, byte3, addr)
			case opcode.BMI    : return instrREL("bmi", byte2, byte3, addr)
			case opcode.BVC    : return instrREL("bvc", byte2, byte3, addr)
			case opcode.BVS    : return instrREL("bvs", byte2, byte3, addr)
			case opcode.BCC    : return instrREL("bcc", byte2, byte3, addr)
			case opcode.BCS    : return instrREL("bcs", byte2, byte3, addr)
			case opcode.BNE    : return instrREL("bne", byte2, byte3, addr)
			case opcode.BEQ    : return instrREL("beq", byte2, byte3, addr)

			case opcode.JSR    : return instrABS("jsr", byte2, byte3)
			case opcode.RTI    : return instrIMP("rti", byte2, byte3)
			case opcode.RTS    : return instrIMP("rts", byte2, byte3)

			case opcode.LDA_IMM: return instrIMM("lda", byte2, byte3)
			case opcode.LDA_ZER: return instrZER("lda", byte2, byte3)
			case opcode.LDA_ZRX: return instrZRX("lda", byte2, byte3)
			case opcode.LDA_ABS: return instrABS("lda", byte2, byte3)
			case opcode.LDA_ABX: return instrABX("lda", byte2, byte3)
			case opcode.LDA_ABY: return instrABY("lda", byte2, byte3)
			case opcode.LDA_PTX: return instrPTX("lda", byte2, byte3)
			case opcode.LDA_PTY: return instrPTY("lda", byte2, byte3)

			case opcode.LDX_IMM: return instrIMM("ldx", byte2, byte3)
			case opcode.LDX_ZER: return instrZER("ldx", byte2, byte3)
			case opcode.LDX_ZRY: return instrZRY("ldx", byte2, byte3)
			case opcode.LDX_ABS: return instrABS("ldx", byte2, byte3)
			case opcode.LDX_ABY: return instrABY("ldx", byte2, byte3)

			case opcode.LDY_IMM: return instrIMM("ldy", byte2, byte3)
			case opcode.LDY_ZER: return instrZER("ldy", byte2, byte3)
			case opcode.LDY_ZRX: return instrZRX("ldy", byte2, byte3)
			case opcode.LDY_ABS: return instrABS("ldy", byte2, byte3)
			case opcode.LDY_ABX: return instrABX("ldy", byte2, byte3)

			case opcode.STA_ZER: return instrZER("sta", byte2, byte3)
			case opcode.STA_ZRX: return instrZRX("sta", byte2, byte3)
			case opcode.STA_ABS: return instrABS("sta", byte2, byte3)
			case opcode.STA_ABX: return instrABX("sta", byte2, byte3)
			case opcode.STA_ABY: return instrABY("sta", byte2, byte3)
			case opcode.STA_PTX: return instrPTX("sta", byte2, byte3)
			case opcode.STA_PTY: return instrPTY("sta", byte2, byte3)

			case opcode.STX_ZER: return instrZER("stx", byte2, byte3)
			case opcode.STX_ZRY: return instrZRY("stx", byte2, byte3)
			case opcode.STX_ABS: return instrABS("stx", byte2, byte3)

			case opcode.STY_ZER: return instrZER("sty", byte2, byte3)
			case opcode.STY_ZRX: return instrZRX("sty", byte2, byte3)
			case opcode.STY_ABS: return instrABS("sty", byte2, byte3)

			case opcode.AND_IMM: return instrIMM("and", byte2, byte3)
			case opcode.AND_ZER: return instrZER("and", byte2, byte3)
			case opcode.AND_ZRX: return instrZRX("and", byte2, byte3)
			case opcode.AND_ABS: return instrABS("and", byte2, byte3)
			case opcode.AND_ABX: return instrABX("and", byte2, byte3)
			case opcode.AND_ABY: return instrABY("and", byte2, byte3)
			case opcode.AND_PTX: return instrPTX("and", byte2, byte3)
			case opcode.AND_PTY: return instrPTY("and", byte2, byte3)

			case opcode.ORA_IMM: return instrIMM("ora", byte2, byte3)
			case opcode.ORA_ZER: return instrZER("ora", byte2, byte3)
			case opcode.ORA_ZRX: return instrZRX("ora", byte2, byte3)
			case opcode.ORA_ABS: return instrABS("ora", byte2, byte3)
			case opcode.ORA_ABX: return instrABX("ora", byte2, byte3)
			case opcode.ORA_ABY: return instrABY("ora", byte2, byte3)
			case opcode.ORA_PTX: return instrPTX("ora", byte2, byte3)
			case opcode.ORA_PTY: return instrPTY("ora", byte2, byte3)

			case opcode.EOR_IMM: return instrIMM("eor", byte2, byte3)
			case opcode.EOR_ZER: return instrZER("eor", byte2, byte3)
			case opcode.EOR_ZRX: return instrZRX("eor", byte2, byte3)
			case opcode.EOR_ABS: return instrABS("eor", byte2, byte3)
			case opcode.EOR_ABX: return instrABX("eor", byte2, byte3)
			case opcode.EOR_ABY: return instrABY("eor", byte2, byte3)
			case opcode.EOR_PTX: return instrPTX("eor", byte2, byte3)
			case opcode.EOR_PTY: return instrPTY("eor", byte2, byte3)

			case opcode.ADC_IMM: return instrIMM("adc", byte2, byte3)
			case opcode.ADC_ZER: return instrZER("adc", byte2, byte3)
			case opcode.ADC_ZRX: return instrZRX("adc", byte2, byte3)
			case opcode.ADC_ABS: return instrABS("adc", byte2, byte3)
			case opcode.ADC_ABX: return instrABX("adc", byte2, byte3)
			case opcode.ADC_ABY: return instrABY("adc", byte2, byte3)
			case opcode.ADC_PTX: return instrPTX("adc", byte2, byte3)
			case opcode.ADC_PTY: return instrPTY("adc", byte2, byte3)

			case opcode.SBC_IMM: return instrIMM("sbc", byte2, byte3)
			case opcode.SBC_ZER: return instrZER("sbc", byte2, byte3)
			case opcode.SBC_ZRX: return instrZRX("sbc", byte2, byte3)
			case opcode.SBC_ABS: return instrABS("sbc", byte2, byte3)
			case opcode.SBC_ABX: return instrABX("sbc", byte2, byte3)
			case opcode.SBC_ABY: return instrABY("sbc", byte2, byte3)
			case opcode.SBC_PTX: return instrPTX("sbc", byte2, byte3)
			case opcode.SBC_PTY: return instrPTY("sbc", byte2, byte3)

			case opcode.CMP_IMM: return instrIMM("cmp", byte2, byte3)
			case opcode.CMP_ZER: return instrZER("cmp", byte2, byte3)
			case opcode.CMP_ZRX: return instrZRX("cmp", byte2, byte3)
			case opcode.CMP_ABS: return instrABS("cmp", byte2, byte3)
			case opcode.CMP_ABX: return instrABX("cmp", byte2, byte3)
			case opcode.CMP_ABY: return instrABY("cmp", byte2, byte3)
			case opcode.CMP_PTX: return instrPTX("cmp", byte2, byte3)
			case opcode.CMP_PTY: return instrPTY("cmp", byte2, byte3)

			case opcode.CPX_IMM: return instrIMM("cpx", byte2, byte3)
			case opcode.CPX_ZER: return instrZER("cpx", byte2, byte3)
			case opcode.CPX_ABS: return instrABS("cpx", byte2, byte3)

			case opcode.CPY_IMM: return instrIMM("cpy", byte2, byte3)
			case opcode.CPY_ZER: return instrZER("cpy", byte2, byte3)
			case opcode.CPY_ABS: return instrABS("cpy", byte2, byte3)

			case opcode.INX    : return instrIMP("inx", byte2, byte3)
			case opcode.DEX    : return instrIMP("dex", byte2, byte3)
			case opcode.INY    : return instrIMP("iny", byte2, byte3)
			case opcode.DEY    : return instrIMP("dey", byte2, byte3)

			case opcode.INC_ZER: return instrZER("inc", byte2, byte3)
			case opcode.INC_ZRX: return instrZRX("inc", byte2, byte3)
			case opcode.INC_ABS: return instrABS("inc", byte2, byte3)
			case opcode.INC_ABX: return instrABX("inc", byte2, byte3)

			case opcode.DEC_ZER: return instrZER("dec", byte2, byte3)
			case opcode.DEC_ZRX: return instrZRX("dec", byte2, byte3)
			case opcode.DEC_ABS: return instrABS("dec", byte2, byte3)
			case opcode.DEC_ABX: return instrABX("dec", byte2, byte3)

			case opcode.ASL_IMP: return instrIMP("asl a", byte2, byte3)
			case opcode.ASL_ZER: return instrZER("asl", byte2, byte3)
			case opcode.ASL_ZRX: return instrZRX("asl", byte2, byte3)
			case opcode.ASL_ABS: return instrABS("asl", byte2, byte3)
			case opcode.ASL_ABX: return instrABX("asl", byte2, byte3)

			case opcode.LSR_IMP: return instrIMP("lsr a", byte2, byte3)
			case opcode.LSR_ZER: return instrZER("lsr", byte2, byte3)
			case opcode.LSR_ZRX: return instrZRX("lsr", byte2, byte3)
			case opcode.LSR_ABS: return instrABS("lsr", byte2, byte3)
			case opcode.LSR_ABX: return instrABX("lsr", byte2, byte3)

			case opcode.ROL_IMP: return instrIMP("rol a", byte2, byte3)
			case opcode.ROL_ZER: return instrZER("rol", byte2, byte3)
			case opcode.ROL_ZRX: return instrZRX("rol", byte2, byte3)
			case opcode.ROL_ABS: return instrABS("rol", byte2, byte3)
			case opcode.ROL_ABX: return instrABX("rol", byte2, byte3)

			case opcode.ROR_IMP: return instrIMP("ror a", byte2, byte3)
			case opcode.ROR_ZER: return instrZER("ror", byte2, byte3)
			case opcode.ROR_ZRX: return instrZRX("ror", byte2, byte3)
			case opcode.ROR_ABS: return instrABS("ror", byte2, byte3)
			case opcode.ROR_ABX: return instrABX("ror", byte2, byte3)

			case opcode.BIT_ZER: return instrZER("bit", byte2, byte3)
			case opcode.BIT_ABS: return instrABS("bit", byte2, byte3)

			case opcode.TXA    : return instrIMP("txa", byte2, byte3)
			case opcode.TAX    : return instrIMP("tax", byte2, byte3)
			case opcode.TYA    : return instrIMP("tya", byte2, byte3)
			case opcode.TAY    : return instrIMP("tay", byte2, byte3)
			case opcode.TXS    : return instrIMP("txs", byte2, byte3)
			case opcode.TSX    : return instrIMP("tsx", byte2, byte3)

			case opcode.CLC    : return instrIMP("clc", byte2, byte3)
			case opcode.SEC    : return instrIMP("sec", byte2, byte3)
			case opcode.CLI    : return instrIMP("cli", byte2, byte3)
			case opcode.SEI    : return instrIMP("sei", byte2, byte3)
			case opcode.CLD    : return instrIMP("cld", byte2, byte3)
			case opcode.SED    : return instrIMP("sed", byte2, byte3)
			case opcode.CLV    : return instrIMP("clv", byte2, byte3)

			case opcode.NOP    : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_2  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_3  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_4  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_5  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_6  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_7  : return instrIMP("nop", byte2, byte3)
			case opcode.NOP_8  : return instrIMP("nop", byte2, byte3)
		}
	}
}


function instrIMP(mnemonic, byte2, byte3)
{
	return mnemonic
}


function instrIMM(mnemonic, byte2, byte3)
{
	return mnemonic + " #$" + byte2.toString(16)
}


function instrZER(mnemonic, byte2, byte3)
{
	return mnemonic + " <$" + byte2.toString(16)
}


function instrZRX(mnemonic, byte2, byte3)
{
	return mnemonic + " <$" + byte2.toString(16) + ", x"
}


function instrZRY(mnemonic, byte2, byte3)
{
	return mnemonic + " <$" + byte2.toString(16) + ", y"
}


function instrABS(mnemonic, byte2, byte3)
{
	return mnemonic + " $" + (byte2 | (byte3 << 8)).toString(16)
}


function instrABX(mnemonic, byte2, byte3)
{
	return mnemonic + " $" + (byte2 | (byte3 << 8)).toString(16) + ", x"
}


function instrABY(mnemonic, byte2, byte3)
{
	return mnemonic + " $" + (byte2 | (byte3 << 8)).toString(16) + ", y"
}


function instrIND(mnemonic, byte2, byte3)
{
	return mnemonic + " ($" + (byte2 | (byte3 << 8)).toString(16) + ")"
}


function instrPTX(mnemonic, byte2, byte3)
{
	return mnemonic + " (<$" + byte2.toString(16) + ", x)"
}


function instrPTY(mnemonic, byte2, byte3)
{
	return mnemonic + " (<$" + byte2.toString(16) + "), y"
}


function instrREL(mnemonic, byte2, byte3, addr)
{
	const signedOffset = ((byte2 & 0x80) == 0) ? byte2 : -(256 - byte2)
	const finalAddr = (addr + 0x10000 + signedOffset) & 0xffff
	
	return mnemonic + " $" + finalAddr.toString(16)
}

File: rust/src/cpu_dis.rs
use cpu_opcodes;


pub fn disassemble_instruction(addr: u16, byte1: u8, byte2: u8, byte3: u8) -> String
{
	match byte1
	{
		cpu_opcodes::BRK     => dis_instr_imp("brk", byte2, byte3),
                             
		cpu_opcodes::PLA     => dis_instr_imp("pla", byte2, byte3),
		cpu_opcodes::PLP     => dis_instr_imp("plp", byte2, byte3),
		cpu_opcodes::PHA     => dis_instr_imp("pha", byte2, byte3),
		cpu_opcodes::PHP     => dis_instr_imp("php", byte2, byte3),
                             
		cpu_opcodes::JMP_ABS => dis_instr_abs("jmp", byte2, byte3),
		cpu_opcodes::JMP_IND => dis_instr_ind("jmp", byte2, byte3),
		cpu_opcodes::BPL     => dis_instr_rel("bpl", byte2, byte3, addr),
		cpu_opcodes::BMI     => dis_instr_rel("bmi", byte2, byte3, addr),
		cpu_opcodes::BVC     => dis_instr_rel("bvc", byte2, byte3, addr),
		cpu_opcodes::BVS     => dis_instr_rel("bvs", byte2, byte3, addr),
		cpu_opcodes::BCC     => dis_instr_rel("bcc", byte2, byte3, addr),
		cpu_opcodes::BCS     => dis_instr_rel("bcs", byte2, byte3, addr),
		cpu_opcodes::BNE     => dis_instr_rel("bne", byte2, byte3, addr),
		cpu_opcodes::BEQ     => dis_instr_rel("beq", byte2, byte3, addr),
                             
		cpu_opcodes::JSR     => dis_instr_abs("jsr", byte2, byte3),
		cpu_opcodes::RTI     => dis_instr_imp("rti", byte2, byte3),
		cpu_opcodes::RTS     => dis_instr_imp("rts", byte2, byte3),
                             
		cpu_opcodes::LDA_IMM => dis_instr_imm("lda", byte2, byte3),
		cpu_opcodes::LDA_ZER => dis_instr_zer("lda", byte2, byte3),
		cpu_opcodes::LDA_ZRX => dis_instr_zrx("lda", byte2, byte3),
		cpu_opcodes::LDA_ABS => dis_instr_abs("lda", byte2, byte3),
		cpu_opcodes::LDA_ABX => dis_instr_abx("lda", byte2, byte3),
		cpu_opcodes::LDA_ABY => dis_instr_aby("lda", byte2, byte3),
		cpu_opcodes::LDA_PTX => dis_instr_ptx("lda", byte2, byte3),
		cpu_opcodes::LDA_PTY => dis_instr_pty("lda", byte2, byte3),
                             
		cpu_opcodes::LDX_IMM => dis_instr_imm("ldx", byte2, byte3),
		cpu_opcodes::LDX_ZER => dis_instr_zer("ldx", byte2, byte3),
		cpu_opcodes::LDX_ZRY => dis_instr_zry("ldx", byte2, byte3),
		cpu_opcodes::LDX_ABS => dis_instr_abs("ldx", byte2, byte3),
		cpu_opcodes::LDX_ABY => dis_instr_aby("ldx", byte2, byte3),
                             
		cpu_opcodes::LDY_IMM => dis_instr_imm("ldy", byte2, byte3),
		cpu_opcodes::LDY_ZER => dis_instr_zer("ldy", byte2, byte3),
		cpu_opcodes::LDY_ZRX => dis_instr_zrx("ldy", byte2, byte3),
		cpu_opcodes::LDY_ABS => dis_instr_abs("ldy", byte2, byte3),
		cpu_opcodes::LDY_ABX => dis_instr_abx("ldy", byte2, byte3),
                             
		cpu_opcodes::STA_ZER => dis_instr_zer("sta", byte2, byte3),
		cpu_opcodes::STA_ZRX => dis_instr_zrx("sta", byte2, byte3),
		cpu_opcodes::STA_ABS => dis_instr_abs("sta", byte2, byte3),
		cpu_opcodes::STA_ABX => dis_instr_abx("sta", byte2, byte3),
		cpu_opcodes::STA_ABY => dis_instr_aby("sta", byte2, byte3),
		cpu_opcodes::STA_PTX => dis_instr_ptx("sta", byte2, byte3),
		cpu_opcodes::STA_PTY => dis_instr_pty("sta", byte2, byte3),
                             
		cpu_opcodes::STX_ZER => dis_instr_zer("stx", byte2, byte3),
		cpu_opcodes::STX_ZRY => dis_instr_zry("stx", byte2, byte3),
		cpu_opcodes::STX_ABS => dis_instr_abs("stx", byte2, byte3),
                             
		cpu_opcodes::STY_ZER => dis_instr_zer("sty", byte2, byte3),
		cpu_opcodes::STY_ZRX => dis_instr_zrx("sty", byte2, byte3),
		cpu_opcodes::STY_ABS => dis_instr_abs("sty", byte2, byte3),
                             
		cpu_opcodes::AND_IMM => dis_instr_imm("and", byte2, byte3),
		cpu_opcodes::AND_ZER => dis_instr_zer("and", byte2, byte3),
		cpu_opcodes::AND_ZRX => dis_instr_zrx("and", byte2, byte3),
		cpu_opcodes::AND_ABS => dis_instr_abs("and", byte2, byte3),
		cpu_opcodes::AND_ABX => dis_instr_abx("and", byte2, byte3),
		cpu_opcodes::AND_ABY => dis_instr_aby("and", byte2, byte3),
		cpu_opcodes::AND_PTX => dis_instr_ptx("and", byte2, byte3),
		cpu_opcodes::AND_PTY => dis_instr_pty("and", byte2, byte3),
                             
		cpu_opcodes::ORA_IMM => dis_instr_imm("ora", byte2, byte3),
		cpu_opcodes::ORA_ZER => dis_instr_zer("ora", byte2, byte3),
		cpu_opcodes::ORA_ZRX => dis_instr_zrx("ora", byte2, byte3),
		cpu_opcodes::ORA_ABS => dis_instr_abs("ora", byte2, byte3),
		cpu_opcodes::ORA_ABX => dis_instr_abx("ora", byte2, byte3),
		cpu_opcodes::ORA_ABY => dis_instr_aby("ora", byte2, byte3),
		cpu_opcodes::ORA_PTX => dis_instr_ptx("ora", byte2, byte3),
		cpu_opcodes::ORA_PTY => dis_instr_pty("ora", byte2, byte3),
                             
		cpu_opcodes::EOR_IMM => dis_instr_imm("eor", byte2, byte3),
		cpu_opcodes::EOR_ZER => dis_instr_zer("eor", byte2, byte3),
		cpu_opcodes::EOR_ZRX => dis_instr_zrx("eor", byte2, byte3),
		cpu_opcodes::EOR_ABS => dis_instr_abs("eor", byte2, byte3),
		cpu_opcodes::EOR_ABX => dis_instr_abx("eor", byte2, byte3),
		cpu_opcodes::EOR_ABY => dis_instr_aby("eor", byte2, byte3),
		cpu_opcodes::EOR_PTX => dis_instr_ptx("eor", byte2, byte3),
		cpu_opcodes::EOR_PTY => dis_instr_pty("eor", byte2, byte3),
                             
		cpu_opcodes::ADC_IMM => dis_instr_imm("adc", byte2, byte3),
		cpu_opcodes::ADC_ZER => dis_instr_zer("adc", byte2, byte3),
		cpu_opcodes::ADC_ZRX => dis_instr_zrx("adc", byte2, byte3),
		cpu_opcodes::ADC_ABS => dis_instr_abs("adc", byte2, byte3),
		cpu_opcodes::ADC_ABX => dis_instr_abx("adc", byte2, byte3),
		cpu_opcodes::ADC_ABY => dis_instr_aby("adc", byte2, byte3),
		cpu_opcodes::ADC_PTX => dis_instr_ptx("adc", byte2, byte3),
		cpu_opcodes::ADC_PTY => dis_instr_pty("adc", byte2, byte3),
                             
		cpu_opcodes::SBC_IMM => dis_instr_imm("sbc", byte2, byte3),
		cpu_opcodes::SBC_ZER => dis_instr_zer("sbc", byte2, byte3),
		cpu_opcodes::SBC_ZRX => dis_instr_zrx("sbc", byte2, byte3),
		cpu_opcodes::SBC_ABS => dis_instr_abs("sbc", byte2, byte3),
		cpu_opcodes::SBC_ABX => dis_instr_abx("sbc", byte2, byte3),
		cpu_opcodes::SBC_ABY => dis_instr_aby("sbc", byte2, byte3),
		cpu_opcodes::SBC_PTX => dis_instr_ptx("sbc", byte2, byte3),
		cpu_opcodes::SBC_PTY => dis_instr_pty("sbc", byte2, byte3),
                             
		cpu_opcodes::CMP_IMM => dis_instr_imm("cmp", byte2, byte3),
		cpu_opcodes::CMP_ZER => dis_instr_zer("cmp", byte2, byte3),
		cpu_opcodes::CMP_ZRX => dis_instr_zrx("cmp", byte2, byte3),
		cpu_opcodes::CMP_ABS => dis_instr_abs("cmp", byte2, byte3),
		cpu_opcodes::CMP_ABX => dis_instr_abx("cmp", byte2, byte3),
		cpu_opcodes::CMP_ABY => dis_instr_aby("cmp", byte2, byte3),
		cpu_opcodes::CMP_PTX => dis_instr_ptx("cmp", byte2, byte3),
		cpu_opcodes::CMP_PTY => dis_instr_pty("cmp", byte2, byte3),
                             
		cpu_opcodes::CPX_IMM => dis_instr_imm("cpx", byte2, byte3),
		cpu_opcodes::CPX_ZER => dis_instr_zer("cpx", byte2, byte3),
		cpu_opcodes::CPX_ABS => dis_instr_abs("cpx", byte2, byte3),
                             
		cpu_opcodes::CPY_IMM => dis_instr_imm("cpy", byte2, byte3),
		cpu_opcodes::CPY_ZER => dis_instr_zer("cpy", byte2, byte3),
		cpu_opcodes::CPY_ABS => dis_instr_abs("cpy", byte2, byte3),
                             
		cpu_opcodes::INX     => dis_instr_imp("inx", byte2, byte3),
		cpu_opcodes::DEX     => dis_instr_imp("dex", byte2, byte3),
		cpu_opcodes::INY     => dis_instr_imp("iny", byte2, byte3),
		cpu_opcodes::DEY     => dis_instr_imp("dey", byte2, byte3),
                             
		cpu_opcodes::INC_ZER => dis_instr_zer("inc", byte2, byte3),
		cpu_opcodes::INC_ZRX => dis_instr_zrx("inc", byte2, byte3),
		cpu_opcodes::INC_ABS => dis_instr_abs("inc", byte2, byte3),
		cpu_opcodes::INC_ABX => dis_instr_abx("inc", byte2, byte3),
                             
		cpu_opcodes::DEC_ZER => dis_instr_zer("dec", byte2, byte3),
		cpu_opcodes::DEC_ZRX => dis_instr_zrx("dec", byte2, byte3),
		cpu_opcodes::DEC_ABS => dis_instr_abs("dec", byte2, byte3),
		cpu_opcodes::DEC_ABX => dis_instr_abx("dec", byte2, byte3),
                             
		cpu_opcodes::ASL_IMP => dis_instr_imp("asl a", byte2, byte3),
		cpu_opcodes::ASL_ZER => dis_instr_zer("asl", byte2, byte3),
		cpu_opcodes::ASL_ZRX => dis_instr_zrx("asl", byte2, byte3),
		cpu_opcodes::ASL_ABS => dis_instr_abs("asl", byte2, byte3),
		cpu_opcodes::ASL_ABX => dis_instr_abx("asl", byte2, byte3),
                             
		cpu_opcodes::LSR_IMP => dis_instr_imp("lsr a", byte2, byte3),
		cpu_opcodes::LSR_ZER => dis_instr_zer("lsr", byte2, byte3),
		cpu_opcodes::LSR_ZRX => dis_instr_zrx("lsr", byte2, byte3),
		cpu_opcodes::LSR_ABS => dis_instr_abs("lsr", byte2, byte3),
		cpu_opcodes::LSR_ABX => dis_instr_abx("lsr", byte2, byte3),
                             
		cpu_opcodes::ROL_IMP => dis_instr_imp("rol a", byte2, byte3),
		cpu_opcodes::ROL_ZER => dis_instr_zer("rol", byte2, byte3),
		cpu_opcodes::ROL_ZRX => dis_instr_zrx("rol", byte2, byte3),
		cpu_opcodes::ROL_ABS => dis_instr_abs("rol", byte2, byte3),
		cpu_opcodes::ROL_ABX => dis_instr_abx("rol", byte2, byte3),
                             
		cpu_opcodes::ROR_IMP => dis_instr_imp("ror a", byte2, byte3),
		cpu_opcodes::ROR_ZER => dis_instr_zer("ror", byte2, byte3),
		cpu_opcodes::ROR_ZRX => dis_instr_zrx("ror", byte2, byte3),
		cpu_opcodes::ROR_ABS => dis_instr_abs("ror", byte2, byte3),
		cpu_opcodes::ROR_ABX => dis_instr_abx("ror", byte2, byte3),
                             
		cpu_opcodes::BIT_ZER => dis_instr_zer("bit", byte2, byte3),
		cpu_opcodes::BIT_ABS => dis_instr_abs("bit", byte2, byte3),
                             
		cpu_opcodes::TXA     => dis_instr_imp("txa", byte2, byte3),
		cpu_opcodes::TAX     => dis_instr_imp("tax", byte2, byte3),
		cpu_opcodes::TYA     => dis_instr_imp("tya", byte2, byte3),
		cpu_opcodes::TAY     => dis_instr_imp("tay", byte2, byte3),
		cpu_opcodes::TXS     => dis_instr_imp("txs", byte2, byte3),
		cpu_opcodes::TSX     => dis_instr_imp("tsx", byte2, byte3),
                             
		cpu_opcodes::CLC     => dis_instr_imp("clc", byte2, byte3),
		cpu_opcodes::SEC     => dis_instr_imp("sec", byte2, byte3),
		cpu_opcodes::CLI     => dis_instr_imp("cli", byte2, byte3),
		cpu_opcodes::SEI     => dis_instr_imp("sei", byte2, byte3),
		cpu_opcodes::CLD     => dis_instr_imp("cld", byte2, byte3),
		cpu_opcodes::SED     => dis_instr_imp("sed", byte2, byte3),
		cpu_opcodes::CLV     => dis_instr_imp("clv", byte2, byte3),
                             
		cpu_opcodes::NOP     => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_2   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_3   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_4   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_5   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_6   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_7   => dis_instr_imp("nop", byte2, byte3),
		cpu_opcodes::NOP_8   => dis_instr_imp("nop", byte2, byte3),
		
		_ => "???".to_string()
	}
}


fn dis_instr_imp(mnemonic: &str, _byte2: u8, _byte3: u8) -> String
{
	mnemonic.to_string()
}


fn dis_instr_imm(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} #${:02x}", mnemonic, byte2)
}


fn dis_instr_zer(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} <${:02x}", mnemonic, byte2)
}


fn dis_instr_zrx(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} <${:02x}", mnemonic, byte2)
}


fn dis_instr_zry(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} <${:02x}, y", mnemonic, byte2)
}


fn dis_instr_abs(mnemonic: &str, byte2: u8, byte3: u8) -> String
{
	format!("{} ${:02x}", mnemonic, byte2 as u16 | ((byte3 as u16) << 8))
}


fn dis_instr_abx(mnemonic: &str, byte2: u8, byte3: u8) -> String
{
	format!("{} ${:02x}, x", mnemonic, byte2 as u16 | ((byte3 as u16) << 8))
}


fn dis_instr_aby(mnemonic: &str, byte2: u8, byte3: u8) -> String
{
	format!("{} ${:02x}, y", mnemonic, byte2 as u16 | ((byte3 as u16) << 8))
}


fn dis_instr_ind(mnemonic: &str, byte2: u8, byte3: u8) -> String
{
	format!("{} (${:02x})", mnemonic, byte2 as u16 | ((byte3 as u16) << 8))
}


fn dis_instr_ptx(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} (<${:02x}, x)", mnemonic, byte2)
}


fn dis_instr_pty(mnemonic: &str, byte2: u8, _byte3: u8) -> String
{
	format!("{} (<${:02x}), y", mnemonic, byte2)
}


fn dis_instr_rel(mnemonic: &str, byte2: u8, _byte3: u8, addr: u16) -> String
{
	let extended_offset = if (byte2 & 0x80) == 0 { byte2 as u16 } else { 0xff00 | byte2 as u16 };
	let final_addr = addr.wrapping_add(extended_offset);
	
	format!("{} ${:04x}", mnemonic, final_addr)
}